File: data/frama_c_problems/mult.c
#include <stdio.h>
#include <limits.h>

/*@
    requires a >= 0 && b >= 0;
    requires a*b <= INT_MAX;
    ensures \result == a*b;
    assigns \nothing;
*/
int mul(int a, int b) {
    int x = a, prod = 0;
    /*@ 
        loop invariant x >= 0;
        loop invariant prod == (a-x)*b;
        loop assigns prod, x;
        loop variant x;
    */
    while(x > 0) {
        //@ assert x >= 1 && prod == (a-x)*b;
        //@ assert prod <= a*b - b;
        prod = prod + b;
        x--;
    }
    return prod;
}

int main() {
    int pdt = mul(2, 5);
    //@ assert pdt == 10;
}

File: transpiled_rust/frama_c_problems/mult.rs
fn mul(a: i32, b: i32) -> i32 {
    let mut x: i32 = a;
    let mut prod: i32 = 0;
    while x > 0 {
        // verus_assert(1);
        // verus_assert(2);
        prod = prod + b;
        x -= 1;
    }
    prod
}

fn main() {
    let pdt: i32 = mul(2, 5);
    // verus_assert(3);
}